src/librustc_mir/transform/validate.rs

```diff
@@ -40,6 +40,40 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
     }
 }
 
+/// Check if src can be assigned into dest.
+/// This is not precise, it will accept some incorrect assignments.
+fn mir_assign_valid_types<'tcx>(tcx: TyCtxt<'tcx>, src: Ty<'tcx>, dest: Ty<'tcx>) -> bool {
+    if src == dest {
+        // Equal types, all is good.
+        return true;
+    }
+
+    // Type-changing assignments can happen for (at least) two reasons:
+    // 1. `&mut T` -> `&T` gets optimized from a reborrow to a mere assignment.
+    // 2. Subtyping is used. While all normal lifetimes are erased, higher-ranked types
+    //    with their late-bound lifetimes are still around and can lead to type differences.
+    // Normalize both of them away.
+    // FIXME: Share this code with `interpret/eval_context.rs`.
+    let normalize = |ty: Ty<'tcx>| {
+        ty.fold_with(&mut BottomUpFolder {
```

Comment:
Btw you can make this cheaper by implementing `TypeFolder` instead, which is only a few lines longer, and doing early returns in the type-visiting method based on flags (which are precomputed so they're basically free) that tell you if there are any erasable lifetimes or normalizable consts (for example, you could choose more things to do/look for).

Although you still have that reference "normalization" aspect, hmpf.

Reply:
I understood maybe half of that so I think I'll leave this to someone else...